File: 2023/day6/main.py
import re
import math


def main():
    with open("input.txt") as f:
        times = next(f)
        distances = next(f)
        times = re.findall(r"\d+", times)
        distances = re.findall(r"\d+", distances)

    s1 = 1
    for time, distance in zip(
        (int(t) for t in times), (int(d) for d in distances)
    ):
        s1 *= number_of_ways_to_beat(time, distance)
    print(f"{s1=}")

    s2 = number_of_ways_to_beat(int(''.join(times)), int(''.join(distances)))
    print(f"{s2=}")


def number_of_ways_to_beat(time, distance):
    speed_to_beat = calculate_speed(time, distance)
    start = math.floor(speed_to_beat+1)
    ways = 0

    for i in range(start, time // 2 + 1):
        current_distance = (time - i) * i
        if current_distance > distance:
            if time - i != i:
                ways += 2
            else:
                ways += 1
        else:
            break

    return ways


def calculate_speed(time, distance):
    """Calculates the speed to have achieved that distance
    considering the time constraint."""
    # base on quadratic equation (-b +- sqrt(b**2-4ac)) / 2
    solution = (time - math.sqrt(time**2-4*distance)) / 2
    return solution


if __name__ == '__main__':
    main()

    # assert number_of_ways_to_beat(7, 9) == 4
    # assert number_of_ways_to_beat(15, 40) == 8
    # assert number_of_ways_to_beat(30, 200) == 9
    # assert number_of_ways_to_beat(71530, 940200) == 71503

File: 2023/day6/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"regexp"
	"strconv"
)

func main() {
	f, err := os.Open("./input.txt")
	defer f.Close()
	if err != nil {
		log.Fatal(err)
	}
	var timesAndDistances [2][]int
	scanner := bufio.NewScanner(f)
	lineNumber := 0
	for scanner.Scan() {
		line := scanner.Text()
		timesAndDistances[lineNumber] = extractNumbers(line)
		lineNumber += 1
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	s1 := 1
	for i := 0; i < len(timesAndDistances[0]); i++ {
		time := timesAndDistances[0][i]
		distance := timesAndDistances[1][i]
		speed := calculateSpeed(float64(time), float64(distance))
		s1 *= numberOfWaysToBeat(time, distance, int(speed)+1)
	}
	fmt.Printf("s1=%d", s1)
}

func numberOfWaysToBeat(time int, distanceToBeat int, speedOfCurrentRecord int) int {
	count := 0
	for i := speedOfCurrentRecord; i <= time/2; i++ {
		currentDistance := i * (time - i)
		if currentDistance > distanceToBeat {
			if time-i != i {
				count += 2
			} else {
				count += 1
			}
		} else {
			break
		}
	}
	return count
}

func extractNumbers(line string) []int {
	var numbers []int
	re := regexp.MustCompile(`\d+`)
	for _, s := range re.FindAllString(line, -1) {
		if n, err := strconv.Atoi(s); err != nil {
			log.Fatal(err)
		} else {
			numbers = append(numbers, n)
		}
	}
	return numbers
}

func calculateSpeed(time, distance float64) float64 {
	return (time - math.Sqrt(math.Pow(time, 2)-4*distance)) / 2
}
